ssl/ssl_lib.c

```diff
@@ -4074,45 +4070,138 @@ int SSL_CTX_set_ct_validation_callback(SSL_CTX *ctx, ct_validation_cb callback,
+        SSLerr(SSL_F_SSL_CT_CB, SSL_R_NO_VALID_SCTS);
+        break;
+    }
+    return 0;
+}
+
+int SSL_CTX_enable_ct(SSL_CTX *ctx, int vtype)
+{
+    void *arg = ssl_ct_arg(vtype);
+
+    if (arg == NULL)
+        return 0;
+    return SSL_CTX_set_ct_validation_callback(ctx, ssl_ct_cb, arg);
+}
+
+int SSL_enable_ct(SSL *s, int vtype)
+{
+    void *arg = ssl_ct_arg(vtype);
+
+    if (arg == NULL)
+        return 0;
+    return SSL_set_ct_validation_callback(s, ssl_ct_cb, arg);
 }
 
 int ssl_validate_ct(SSL *s)
 {
     int ret = 0;
     X509 *cert = s->session != NULL ? s->session->peer : NULL;
-    X509 *issuer = NULL;
+    X509 *issuer;
+    struct dane_st *dane = &s->dane;
     CT_POLICY_EVAL_CTX *ctx = NULL;
     const STACK_OF(SCT) *scts;
 
-    /* If no callback is set, attempt no validation - just return success */
-    if (s->ct_validation_callback == NULL)
+    /*
+     * If no callback is set, the peer is anonymous, or its chain is invalid,
+     * skip SCT callbacks - just return success.  Note, that SCT internal
+     * processing has already taken place, just by virtue of the fact that a
+     * callback is defined, the only reason to issue the callback is to give
+     * the application the opportunity to inspect the SCTs early and
+     * potentially abort the connection early.
+     *
+     * Applications can still check the processed SCTs at handshake completion.
+     *
+     * In applications that continue with unverified peers, SCT is the least of
+     * their concerns.  It makes little sense to process SCTs for an invalid
+     * chain, and might lead to greater false trust of an invalid peer.
+     * Clients what disable chain verification (continue despite failure) are
+     * not further confused with potentially misleading SCTs from unverified
+     * chains.
+     *
+     * Chains consisting of just a directly trusted leaf certificate (this case
+     * overlaps with DANE-EE(3)) don't have an issuer CA and so certificate
+     * transparency is again out of scope.
+     *
+     * The above exclusions notwithstanding the vast majority of non-DANE peers
+     * will have rather ordinary certificate chains validated by typical
+     * applications that perform certificate verification and therefore will
+     * process SCTs when enabled.
+     */
+    if (s->ct_validation_callback == NULL || cert == NULL ||
+        s->verify_result != X509_V_OK ||
+        s->verified_chain == NULL ||
+        sk_X509_num(s->verified_chain) <= 1)
         return 1;
 
-    if (cert == NULL) {
-        SSLerr(SSL_F_SSL_VALIDATE_CT, SSL_R_NO_CERTIFICATE_ASSIGNED);
-        goto end;
-    }
-
-    if (s->verified_chain != NULL && sk_X509_num(s->verified_chain) > 1)
-        issuer = sk_X509_value(s->verified_chain, 1);
+    /*
+     * CT not applicable for chains validated via DANE-TA(2) or DANE-EE(3)
+     * trust-anchors.  See https://tools.ietf.org/html/rfc7671#section-4.2
+     */
+    if (DANETLS_ENABLED(dane) && dane->mtlsa != NULL)
```

Comment:
Nit: I'd prefer this if statement have braces, given the length of the switch it contains, but the style guide doesn't require it so feel free to ignore this.